File: parad/parad/crypto.py
"""AES-256-CBC file encryption for parad local database."""

import os
import threading
import hashlib
from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from cryptography.hazmat.primitives import hashes

SALT = b"paradox-salt"
KDF_ITERATIONS = 256_000
KEY_LENGTH = 32  # 256 bits
IV_LENGTH = 16  # 128 bits

# Every SQLite database starts with this 16-byte magic header.
SQLITE_MAGIC = b"SQLite format 3\x00"

# Every ciphertext block is 16 bytes; the IV is 16 bytes, so a valid
# encrypted blob is always IV + at least one AES block.
MIN_ENCRYPTED_LENGTH = IV_LENGTH + 16


class DecryptionError(ValueError):
    """Raised when decryption fails.

    Covers wrong passphrase, truncated/corrupt data, invalid PKCS7
    padding, and payloads that are not SQLite databases.  Subclasses
    ValueError so existing ``except ValueError`` handlers keep working.
    """

    DEFAULT_MESSAGE = "Invalid passphrase or corrupt database file"


# Memoized derived keys.  PBKDF2 is deterministic and deliberately slow
# (256k HMAC-SHA512 iterations), and the engine re-encrypts on every
# close and re-decrypts on every open — recomputing the key each time is
# wasteful.  The cache is in-memory only and never persisted; the derived
# bytes are identical, so the on-disk format is unchanged.
_KEY_CACHE: dict[str, bytes] = {}
_KEY_CACHE_LOCK = threading.Lock()


def derive_key(passphrase: str) -> bytes:
    """Derive a 256-bit key from passphrase using PBKDF2-HMAC-SHA512."""
    cached = _KEY_CACHE.get(passphrase)
    if cached is not None:
        return cached
    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA512(),
        length=KEY_LENGTH,
        salt=SALT,
        iterations=KDF_ITERATIONS,
    )
    key = kdf.derive(passphrase.encode("utf-8"))
    with _KEY_CACHE_LOCK:
        _KEY_CACHE[passphrase] = key
    return key


def encrypt_file(data: bytes, passphrase: str) -> bytes:
    """Encrypt bytes with AES-256-CBC. Returns IV + ciphertext."""
    key = derive_key(passphrase)
    iv = os.urandom(IV_LENGTH)
    cipher = Cipher(algorithms.AES(key), modes.CBC(iv))
    encryptor = cipher.encryptor()
    # PKCS7 padding
    pad_len = 16 - (len(data) % 16)
    padded = data + bytes([pad_len] * pad_len)
    ciphertext = encryptor.update(padded) + encryptor.finalize()
    return iv + ciphertext


def decrypt_file(data: bytes, passphrase: str) -> bytes:
    """Decrypt AES-256-CBC data (IV + ciphertext). Returns original bytes.

    Raises :class:`DecryptionError` when the data cannot be decrypted:
    wrong passphrase, truncated/corrupt input, invalid PKCS7 padding, or
    a payload that does not start with the SQLite magic header.
    """
    key = derive_key(passphrase)
    if len(data) < MIN_ENCRYPTED_LENGTH:
        raise DecryptionError(
            f"{DecryptionError.DEFAULT_MESSAGE}: data too short "
            f"({len(data)} bytes, need at least {MIN_ENCRYPTED_LENGTH})"
        )
    iv = data[:IV_LENGTH]
    ciphertext = data[IV_LENGTH:]
    if len(ciphertext) % 16 != 0:
        raise DecryptionError(
            f"{DecryptionError.DEFAULT_MESSAGE}: ciphertext is not "
            "block-aligned"
        )
    cipher = Cipher(algorithms.AES(key), modes.CBC(iv))
    decryptor = cipher.decryptor()
    padded = decryptor.update(ciphertext) + decryptor.finalize()
    # Strict PKCS7 validation: pad_len must be 1..16 and every trailing
    # byte must equal pad_len.
    pad_len = padded[-1]
    if pad_len < 1 or pad_len > 16:
        raise DecryptionError(
            f"{DecryptionError.DEFAULT_MESSAGE}: invalid PKCS7 padding "
            f"(pad byte {pad_len})"
        )
    if padded[-pad_len:] != bytes([pad_len]) * pad_len:
        raise DecryptionError(
            f"{DecryptionError.DEFAULT_MESSAGE}: invalid PKCS7 padding"
        )
    plaintext = padded[:-pad_len]
    if not plaintext.startswith(SQLITE_MAGIC):
        raise DecryptionError(
            f"{DecryptionError.DEFAULT_MESSAGE}: not a SQLite database"
        )
    return plaintext


def validate_passphrase(data: bytes, passphrase: str) -> bool:
    """Return True if *data* decrypts with *passphrase*, False otherwise.

    Never raises — useful for a ``status``/``ping`` style command that
    just wants a yes/no answer about whether a passphrase matches.
    """
    try:
        decrypt_file(data, passphrase)
        return True
    except (DecryptionError, ValueError, IndexError, OSError):
        return False

File: client/src/crypto.ts
import { createCipheriv, createDecipheriv, pbkdf2Sync, randomBytes } from 'node:crypto';
import { DecryptionError } from './errors.js';

export { DecryptionError } from './errors.js';

export const SALT = Buffer.from('paradox-salt', 'utf-8');
export const KDF_ITERATIONS = 256_000;
export const KEY_LENGTH = 32; // 256 bits
export const IV_LENGTH = 16; // 128 bits

// Every SQLite database starts with this 16-byte magic header.
export const SQLITE_MAGIC = Buffer.from('SQLite format 3\x00', 'utf-8');

// Every ciphertext block is 16 bytes; the IV is 16 bytes, so a valid
// encrypted blob is always IV + at least one AES block.
export const MIN_ENCRYPTED_LENGTH = IV_LENGTH + 16;

// Memoized derived keys. PBKDF2 is deterministic and deliberately slow
// (256k HMAC-SHA512 iterations), and the engine re-encrypts on every
// close and re-decrypts on every open. The cache is in-memory only.
const keyCache = new Map<string, Buffer>();

export function deriveKey(passphrase: string): Buffer {
  const cached = keyCache.get(passphrase);
  if (cached) return cached;
  const key = pbkdf2Sync(passphrase, SALT, KDF_ITERATIONS, KEY_LENGTH, 'sha512');
  keyCache.set(passphrase, key);
  return key;
}

/** Encrypt bytes with AES-256-CBC. Returns IV + ciphertext. */
export function encryptFile(data: Buffer, passphrase: string): Buffer {
  const key = deriveKey(passphrase);
  const iv = randomBytes(IV_LENGTH);
  const padLen = 16 - (data.length % 16);
  const padded = Buffer.concat([data, Buffer.alloc(padLen, padLen)]);
  const cipher = createCipheriv('aes-256-cbc', key, iv).setAutoPadding(false);
  const ciphertext = Buffer.concat([cipher.update(padded), cipher.final()]);
  return Buffer.concat([iv, ciphertext]);
}

/**
 * Decrypt AES-256-CBC data (IV + ciphertext). Returns the original bytes.
 *
 * Throws DecryptionError on wrong passphrase, truncated/corrupt input,
 * invalid PKCS7 padding, or a payload that does not start with the
 * SQLite magic header.
 */
export function decryptFile(data: Buffer, passphrase: string): Buffer {
  const key = deriveKey(passphrase);
  if (data.length < MIN_ENCRYPTED_LENGTH) {
    throw new DecryptionError(
      `${new DecryptionError().message}: data too short (${data.length} bytes, need at least ${MIN_ENCRYPTED_LENGTH})`,
    );
  }
  const iv = data.subarray(0, IV_LENGTH);
  const ciphertext = data.subarray(IV_LENGTH);
  if (ciphertext.length % 16 !== 0) {
    throw new DecryptionError(
      `${new DecryptionError().message}: ciphertext is not block-aligned`,
    );
  }
  const decipher = createDecipheriv('aes-256-cbc', key, iv).setAutoPadding(false);
  const padded = Buffer.concat([decipher.update(ciphertext), decipher.final()]);
  // Strict PKCS7 validation: padLen must be 1..16 and every trailing
  // byte must equal padLen.
  const padLen = padded[padded.length - 1];
  if (padLen < 1 || padLen > 16) {
    throw new DecryptionError(
      `${new DecryptionError().message}: invalid PKCS7 padding (pad byte ${padLen})`,
    );
  }
  const padStart = padded.length - padLen;
  for (let i = padStart; i < padded.length; i++) {
    if (padded[i] !== padLen) {
      throw new DecryptionError(
        `${new DecryptionError().message}: invalid PKCS7 padding`,
      );
    }
  }
  const plaintext = padded.subarray(0, padStart);
  if (plaintext.length < SQLITE_MAGIC.length || !plaintext.subarray(0, SQLITE_MAGIC.length).equals(SQLITE_MAGIC)) {
    throw new DecryptionError(
      `${new DecryptionError().message}: not a SQLite database`,
    );
  }
  return Buffer.from(plaintext);
}

/** Return true if data decrypts with passphrase, false otherwise. Never throws. */
export function validatePassphrase(data: Buffer, passphrase: string): boolean {
  try {
    decryptFile(data, passphrase);
    return true;
  } catch {
    return false;
  }
}
